ext/wmq_queue_manager.c

```diff
@@ -243,8 +246,22 @@ VALUE QueueManager_initialize(VALUE self, VALUE hash)
         WMQ_HASH2MQCHARS(hash,security_user_data,          pmqcd->SecurityUserData)
         WMQ_HASH2MQCHARS(hash,send_user_data,              pmqcd->SendUserData)
         WMQ_HASH2MQCHARS(hash,receive_user_data,           pmqcd->ReceiveUserData)
-        WMQ_HASH2MQCHARS(hash,user_identifier,             pmqcd->UserIdentifier)
-        WMQ_HASH2MQCHARS(hash,password,                    pmqcd->Password)
+
+        /* MQCSP connection authentication with user ID and password */
+        VALUE user_identifier = rb_hash_aref(hash, ID2SYM(ID_user_identifier));
+        VALUE password = rb_hash_aref(hash, ID2SYM(ID_password));
+        if (!NIL_P(user_identifier))
+        {
+            pqm->security_parameters.AuthenticationType = MQCSP_AUTH_USER_ID_AND_PWD;
+            pqm->security_parameters.CSPUserIdPtr = StringValuePtr(user_identifier);
```

Comment:
Do these strings need to be copied and freed, the same way `long_remote_user_id_ptr` is (in this file)? I also noticed that `long_remote_user_id_ptr` gets null terminated, but I'm not sure if that's important.

Reply:
Good point - the GC could run between `initialize` and `new` and remove those Ruby Strings from the heap.

Definitely don't need to null terminate them through - the MQ library takes a length.